Common/Util/BusyCollection.cs

```diff
@@ -0,0 +1,111 @@
+ * limitations under the License.
+ *
+*/
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using QuantConnect.Interfaces;
+
+namespace QuantConnect.Util
+{
+    /// <summary>
+    /// A non blocking <see cref="IBusyCollection{T}"/> implementation
+    /// </summary>
+    /// <typeparam name="T">The item type being processed</typeparam>
+    public class BusyCollection<T> : IBusyCollection<T>
+    {
+        private readonly ConcurrentQueue<T> _collection = new ConcurrentQueue<T>();
+        private readonly ManualResetEventSlim _processingCompletedEvent = new ManualResetEventSlim(true);
+
+        /// <summary>
+        /// Gets a wait handle that can be used to wait until this instance is done
+        /// processing all of it's item
+        /// </summary>
+        public WaitHandle WaitHandle => _processingCompletedEvent.WaitHandle;
+
+        /// <summary>
+        /// Gets the number of items held within this collection
+        /// </summary>
+        public int Count => _collection.Count;
+
+        /// <summary>
+        /// Returns true if processing, false otherwise
+        /// </summary>
+        public bool IsBusy => _collection.Count > 0 || !_processingCompletedEvent.IsSet;
+
+        /// <summary>
+        /// Adds the items to this collection
+        /// </summary>
+        /// <param name="item">The item to be added</param>
+        public void Add(T item)
+        {
+            Add(item, CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Adds the items to this collection
+        /// </summary>
+        /// <param name="item">The item to be added</param>
+        /// <param name="cancellationToken">A cancellation token to observer</param>
+        public void Add(T item, CancellationToken cancellationToken)
+        {
+            // we're adding work to be done, mark us as busy
+            _processingCompletedEvent.Reset();
+            _collection.Enqueue(item);
+        }
+
+        /// <summary>
+        /// Provides a consuming enumerable for items in this collection.
+        /// </summary>
+        /// <returns>An enumerable that removes and returns items from the collection</returns>
+        public IEnumerable<T> GetConsumingEnumerable()
+        {
+            return GetConsumingEnumerable(CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Provides a consuming enumerable for items in this collection.
+        /// </summary>
+        /// <param name="cancellationToken">A cancellation token to observer</param>
+        /// <returns>An enumerable that removes and returns items from the collection</returns>
+        public IEnumerable<T> GetConsumingEnumerable(CancellationToken cancellationToken)
+        {
+            T item;
+            while (_collection.TryDequeue(out item))
+            {
+                yield return item;
+            }
+            _processingCompletedEvent.Set();
```

Comment:
hmm... this is tough. it's small, but there exists a race condition between [enqueuing](https://github.com/QuantConnect/Lean/pull/3184/files?file-filters%5B%5D=.cs&file-filters%5B%5D=.csproj#diff-ef82eecda24fe8fdf993e6c251ad18f7R67) a new piece of data and resetting the handle here.

Consider we're adding an item, we flip the switch to say we're currently processing, then that thread stops running, another thread processing `GetConsumingEnumerable` had just emptied the `_collection` and now sets the completed event but there's still another item in the queue.

I could see something like this to fix it:
```
// replacing impl of Add(T, CancellationToken)
lock (sync)
{
    // we're adding work to be done, mark us as busy
    _processingCompletedEvent.Reset();
    _collection.Enqueue(item);
}

// replacing line 91
lock (sync)
{
    if (!_collection.TryPeek(out var _))
    {
        _processingCompletedEvent.Set();
    }
}
```

All that being said, the surface area of this potential race is exceedingly slim. This would serialize additions to the collection, but if I recall correctly, we're not pumping massive numbers of items into these collections, also, contention on the lock should usually be non-existent.